metabase/metabase
# Consolidate action form logic

Based on #29289 and #29290
There's also a follow up PR that just moves around utilities once again #29328

### Description

Action form logic ended up being spread across a few layers and components. This PR consolidates as much logic as possible inside the `ActionForm` component and a new `useActionForm` hook. That includes:

* cleaning up initial values
* generating form fields out of implicit action parameters
* stripping out unchanged values for implicit update actions
* cleaning up values before submitting a form
* figuring out submit button label props (variant and label are configurable)
* hiding fields that don't need user's input (parameters connected to dashboard filters)

Overall, these changes should make the `ActionForm` the go-to component to use when we need an action form.

You can see how easier it is to add a new action form by looking an changes to the `PublicAction` component:

<details>
<summary>Before</summary>

```tsx
function PublicAction({ action, publicId, onError }: Props) {
  const [isSubmitted, setSubmitted] = useState(false);
  const successMessage = getSuccessMessage(action);

  // Must manually handle implicit actions that don't have form fields
  // Has to be done every time the ActionForm is used
  const formSettings = useMemo(() => {
    const actionSettings = action.visualization_settings || {};
    const fieldSettings =
      actionSettings.fields ||
      generateFieldSettingsFromParameters(action.parameters);
    return {
      ...actionSettings,
      fields: fieldSettings,
    };
  }, [action]);

  const handleSubmit = useCallback(
    async (values: ParametersForActionExecution) => {
      try {
        // Must clean the values before submitting
        // Has to be done every time the ActionForm is used
        const parameters = setNumericValues(values, formSettings.fields);
        await PublicApi.executeAction({ uuid: publicId, parameters });
        setSubmitted(true);
      } catch (error) {
        onError(error as AppErrorDescriptor);
      }
    },
    [publicId, formSettings, onError],
  );

  if (isSubmitted) {
    return <FormResultMessage>{successMessage}</FormResultMessage>;
  }

  return (
    <FormContainer>
      <FormTitle>{action.name}</FormTitle>
      <ActionForm
        parameters={action.parameters}
        formSettings={formSettings}
        onSubmit={handleSubmit}
      />
    </FormContainer>
  );
}
```

</details>

<details>
<summary>After</summary>

```tsx
function PublicAction({ action, publicId, onError }: Props) {
  const [isSubmitted, setSubmitted] = useState(false);
  const successMessage = getSuccessMessage(action);

  const handleSubmit = useCallback(
    async (parameters: ParametersForActionExecution) => {
      try {
        await PublicApi.executeAction({ uuid: publicId, parameters });
        setSubmitted(true);
      } catch (error) {
        onError(error as AppErrorDescriptor);
      }
    },
    [publicId, onError],
  );

  if (isSubmitted) {
    return <FormResultMessage>{successMessage}</FormResultMessage>;
  }

  return (
    <FormContainer>
      <FormTitle>{getFormTitle(action)}</FormTitle>
      <ActionForm action={action} onSubmit={handleSubmit} />
    </FormContainer>
  );
}
```

</details>

**Changes to action/dashboard parameter mapping**

There's a common use-case for actions on dashboards when we don't want to show the full action form, but only parameters not mapped to dashboard filters.

For example, we can have an action doing `UPDATE subscriptions SET plan = {{ plan }} WHERE id = {{ subscription_id }}`. We could add this action to a dashboard called "Subscription Details" with a subscription ID filter and an object detail view on it. In that case, we don't really want people to manually specify the subscription ID, as we can take a value directly from the filter.

Prior to this change, we'd figure out which parameters are missing and pass them to the `ActionForm`. Once the `ActionForm` is submitted, the parent component receives the submitted values through the `onSubmit` prop, merge then with dashboard filter values and sends an HTTP request to run the action. Overall, it looked like that:

<details>
<summary>Snippet</summary>

```jsx
function DashboardActionForm({ action, dashboardFilterValues, runAction }) {
  const missingParameters = getNotProvidedParameters(action.parameters, dashboardFilterValues });

  const fieldSettings = action.visualization_settings.fields || generateFieldSettingsFromParams(action.parameters)

  const handleSubmit = values => {
    runAction({
      ...values,
      ...cleanValues(dashboardFilterValues),
    })
  }

  return (
    <ActionForm
       parameters={missingParameters}
       fieldSettings={fieldSettings}
       onSubmit={handleSubmit}
    />
  )
}
```

</details>

Given this design, the parent component had to ensure that dashboard filter values have expected types (e.g. coerse strings to numbers) which is one more thing an engineer had to keep in mind while working with action forms.

This PR solves this by changing the data flow and adding a `hiddenFields` prop to the `ActionForm`:

<details>
<summary>Snippet</summary>

```jsx
function DashboardActionForm({ action, dashboardFilterValues, runAction }) {
  const missingParameters = getNotProvidedParameters(action.parameters, dashboardFilterValues });
  const hiddenFields = missingParameters.map(parameter => parameter.id)

  return (
    <ActionForm
       action={action}

       // We pass dashboard filter values as initial values here
       // and also instruct the ActionForm to visually hide them.
       // In that way, they end up in the values object when the form is submitted,
       // but still remain hidden for the end user.
       // This approach should be much more reusable
       initialValues={dashboardFilterValues}
       hiddenFields={hiddenFields}

       // As the ActionForm now handles cleaning up values,
       // we no longer need to do this here
       onSubmit={runAction}
    />
  )
}
```

</details>

### How to verify

1. CI should be green
2. Play around with actions and ensure everything works the same as on `master`


- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/metabase/metabase/29326)
